src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+
+                instruction = Instruction::new(Op::TypeVector);
+                instruction.set_result(id);
+                instruction.add_operand(scalar_id);
+                instruction.add_operand(size as u32);
+
+                self.lookup_type.insert(id, handle);
+            }
+            crate::TypeInner::Matrix {
+                columns,
+                rows: _,
+                kind,
+                width,
+            } => {
+                let scalar_handle = self.find_scalar_handle(arena, &kind, &width);
+                let scalar_id = self.get_type_id(arena, &scalar_handle);
+
+                instruction = Instruction::new(Op::TypeMatrix);
+                instruction.set_result(id);
+                instruction.add_operand(scalar_id);
+                instruction.add_operand(columns as u32);
+            }
+            crate::TypeInner::Pointer { base, class } => {
+                let type_id = self.get_type_id(arena, &base);
+                instruction = Instruction::new(Op::TypePointer);
+                instruction.set_result(id);
+                instruction.add_operand(class as u32);
+                instruction.add_operand(type_id);
+
+                self.lookup_type.insert(id, handle);
+            }
+            crate::TypeInner::Array { base, size } => {
+                let type_id = self.get_type_id(arena, &base);
+
+                instruction = Instruction::new(Op::TypeArray);
+                instruction.set_result(id);
+                instruction.add_operand(type_id);
+
+                match size {
+                    crate::ArraySize::Static(word) => {
+                        instruction.add_operand(word);
+                    }
+                    _ => panic!("Array size {:?} unsupported", size),
+                }
+
+                self.lookup_type.insert(id, base);
+            }
+            crate::TypeInner::Struct { ref members } => {
+                instruction = Instruction::new(Op::TypeStruct);
+                instruction.set_result(id);
+
+                for member in members {
+                    let type_id = self.get_type_id(arena, &member.ty);
+                    instruction.add_operand(type_id);
+                }
+
+                self.lookup_type.insert(id, handle);
+            }
+            crate::TypeInner::Image { base, dim, flags } => {
+                let type_id = self.get_type_id(arena, &base);
+                self.try_add_capabilities(dim.required_capabilities());
+
+                instruction = Instruction::new(Op::TypeImage);
+                instruction.set_result(id);
+                instruction.add_operand(type_id);
+                instruction.add_operand(dim as u32);
+
+                // TODO Add Depth, but how to determine? Not yet in the WGSL spec
+                instruction.add_operand(1);
+
+                if flags.contains(crate::ImageFlags::ARRAYED) {
+                    instruction.add_operand(1);
```

Comment:
please prefer to use if as an expression. i.e. `instruction.add_operand(if .. else ..);`

Reply:
Sure, always wanting to adapt more to the idiomatic rust ways. Do you mean something like this:
```rust
instruction.add_operand(if flags.contains(crate::ImageFlags::ARRAYED) {
    1
} else {
    0
});
```
Or how would you want to format this?

Reply:
Yes, that's cool